### PR TITLE
Adds functionality to cleanup last errors on success

### LINKED ISSUE #1215: Remove Last Error when Step has been successfully processed
**What would you like to be added**:
Currently the last error is being displayed until we either hit another error or the cluster has been successfully reconciled. It would be more useful to remove that last error if we passed this with a successful processed step.

**Why is this needed**:
Improves operator productivity.

### PR DESCRIPTION
**Pull request by plkokanov**
**What this PR does / why we need it**:
The flow task executor can now be passed an `ErrorCleaner` function and `ErrorContext` which will hold the IDs of the errors/tasks which caused errors during the previous reconciliation. Currently the task names are used as errorIDs but they could be changed to use sha hashes. The ErrorCleaner will be called when a task which previously had an error now succeeds.

Since errors can be generated from functions outside of the flow execution and should also be cleaned up an `ErrorContext` is used. It contains the IDs of the lastErrors from the previous reconciliation and also tracks the IDs of all errors/tasks of the current reconciliation and causes a panic if more than one error/task share the same ID (otherwise there will be problems when trying to clean them up)

There is now an reconcilationerrorhandler package which can be used for cleaning up errors caused by tasks not executed by the flow.

**Which issue(s) this PR fixes**:
Fixes #1215 

**Special notes for your reviewer**:
The names are a bit long (reconciliationerrorhandler) so I am accepting suggestions on naming and code placements :)

**Release note**:

```improvement operator
There is a new field in the status of `garden.sapcloud.io/v1beta1.Shoot` and `core.gardener.cloud/v1alpha1.Shoot`: `.status.lastErrors[]` contains a list of all errors which occurred during the last operation on the `Shoot` resource. Each error in the list has a unique `TaskID` which depends on the task that caused the error.
```
```improvement operator
Once a step which caused an error during the previous operation on a `Shoot` resource completes successfully,  the corresponding error is removed from `.status.lastErrors[]`.
```

### CODE REVIEW

File: pkg/apis/core/v1alpha1/types_common.go [hunk 1]
```diff
@@ -34,6 +34,9 @@ const (
 type LastError struct {
 	// A human readable message indicating details about the last error.
 	Description string `json:"description"`
+	// Name of the task which caused this last error
+	// +optional
+	TaskID string `json:"taskName,omitempty"`
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/types_common.go]:
```suggestion
	TaskID *string `json:"taskName,omitempty"`
```

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

File: pkg/apis/core/v1alpha1/types_common.go [hunk 2]
```diff
@@ -47,6 +50,11 @@ func (l *LastError) GetDescription() string {
 	return l.Description
 }
 
+// GetTaskID implements LastError
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/types_common.go:53]:
```suggestion
// GetTaskID gets the task ID of the last error. 
```

File: pkg/apis/garden/types_common.go [hunk 3]
```diff
@@ -36,6 +36,9 @@ const (
 type LastError struct {
 	// A human readable message indicating details about the last error.
 	Description string
+	// ID of the task which caused this last error
+	// +optional
```
Comment (vpnachev) [inline pkg/apis/garden/types_common.go]:
```suggestion
```
this is not required in the internal version

File: pkg/apis/garden/types_common.go [hunk 4]
```diff
@@ -36,6 +36,9 @@ const (
...
+	TaskID string
```
Comment (vpnachev) [inline pkg/apis/garden/types_common.go]:
```suggestion
	TaskID *string
```

File: pkg/apis/garden/v1beta1/types.go [hunk 5]
```diff
@@ -723,9 +723,9 @@ type ShootStatus struct {
 	// LastOperation holds information about the last operation on the Shoot.
 	// +optional
 	LastOperation *gardencorev1alpha1.LastOperation `json:"lastOperation,omitempty"`
-	// LastError holds information about the last occurred error during an operation.
+	// LastErrors holds information about the last occurred error(s) during an operation.
 	// +optional
-	LastError *gardencorev1alpha1.LastError `json:"lastError,omitempty"`
+	LastErrors []gardencorev1alpha1.LastError `json:"lastErrors,omitempty"`
```
Comment (vpnachev) [inline pkg/apis/garden/v1beta1/types.go:753]: This change might cause issues....
I mean the API might not start up because it cannot decode existing objects.

**Review by vpnachev — COMMENTED** [review]
Some comments about the API changes.

**Comment by rfranzke** [comment]
Hi @plkokanov, what about this PR? Any plans to make it ready for review soon?

**Comment by plkokanov** [comment]
@rfranzke will be done tommorrow, just need to run a few tests

File: pkg/apis/core/v1alpha1/types_common.go (hunk 1 above)
Comment (plkokanov) [reply pkg/apis/core/v1alpha1/types_common.go]: done

File: pkg/apis/core/v1alpha1/types_common.go (hunk 2 above)
Comment (plkokanov) [reply pkg/apis/core/v1alpha1/types_common.go:53]: Followed the comments on the other similar methods. Maybe we can change them all with another PR.

File: pkg/apis/garden/types_common.go (hunk 4 above)
Comment (plkokanov) [reply pkg/apis/garden/types_common.go]: done

File: pkg/apis/garden/v1beta1/types.go (hunk 5 above)
Comment (plkokanov) [reply pkg/apis/garden/v1beta1/types.go:753]: Testing it locally didn't cause errors. I guess because both are optional the status update simply removes one and adds the other.

File: pkg/apis/garden/types_common.go (hunk 3 above)
Comment (plkokanov) [reply pkg/apis/garden/types_common.go]: Kept this because of the same comment on the Codes []ErrorCode below.

File: pkg/apis/core/v1alpha1/types_common.go [hunk 6]
```diff
@@ -34,6 +34,9 @@ const (
 type LastError struct {
 	// A human readable message indicating details about the last error.
 	Description string `json:"description"`
+	// Name of the task which caused this last error
+	// +optional
+	TaskID *string `json:"taskName,omitempty"`
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/types_common.go]: ### Question
Could you use one name for the serialization and for the variable, e.g.

```suggestion
	TaskID *string `json:"taskID,omitempty"`
```
or 
```suggestion
	TaskName *string `json:"taskName,omitempty"`
```

File: pkg/apis/core/types_common.go [hunk 7]
```diff
@@ -36,6 +36,9 @@ const (
 type LastError struct {
 	// A human readable message indicating details about the last error.
 	Description string
+	// Name of the task which caused this last error
+	// +optional
+	TaskID *string `json:"taskName,omitempty"`
```
Comment (vpnachev) [inline pkg/apis/core/types_common.go]: `+optional` and ``json:"taskName,omitempty"`` are not needed here.

File: pkg/apis/garden/types.go [hunk 8]
```diff
@@ -836,9 +836,10 @@ type ShootStatus struct {
 	// Gardener holds information about the Gardener which last acted on the Shoot.
 	Gardener Gardener
 	// LastOperation holds information about the last operation on the Shoot.
+	// +optional
```
Comment (vpnachev) [inline pkg/apis/garden/types.go:845]: `+optional` is not required in the internal version

File: pkg/apis/garden/v1beta1/types.go (hunk 5 above)
Comment (vpnachev) [reply pkg/apis/garden/v1beta1/types.go:753]: I am reading https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md and still wondering whether this change is acceptable.

There is even example with switching to array.
```golang
// Still API v6.
type Frobber struct {
  Height int           `json:"height"`
  Width  int           `json:"width"`
  Param  string        `json:"param"`  // the first param
  Params []string      `json:"params"` // all of the params
}
```


I think that both fields should be presented in the v1beta1 API with the old one marked as deprecated and removed after couple of releases.
Do you think a conversion between both fields is feasible?
Basically something like this:
```golang
if LastError != nil {
    LastErrors = append(LastErrors, *LastError)
}
```

@mvladev @rfranzke wdyt?

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go [hunk 9]
```diff
@@ -44,98 +45,127 @@ import (
...
-		return gardencorev1alpha1helper.LastError(fmt.Sprintf("Failed to check the deletion timestamp for the Shoot namespace (%s)", err.Error()))
+		return gardencorev1alpha1helper.NewWrappedLastErrors(gardencorev1alpha1helper.FormatLastErrDescription(err), err)
 	}
 
+	// Unregister the Shoot as Seed cluster if it was annotated to be a seed and is in the garden namespace
+
 	// We check whether the kube-apiserver deployment exists in the shoot namespace. If it does not, then we assume
 	// that it has never been deployed successfully, or that we have deleted it in a previous run because we already
 	// cleaned up. We follow that no (more) resources can have been deployed in the shoot cluster, thus there is nothing
 	// to delete anymore.
```
Comment (vpnachev) [inline pkg/controllermanager/controller/shoot/shoot_control_delete.go]: These comments should be (re)moved.

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go [hunk 10]
```diff
@@ -44,98 +45,127 @@ import (
...
-		}
-		kubeAPIServerDeploymentFound = false
-	}
-	if deploymentKubeAPIServer.DeletionTimestamp != nil {
-		kubeAPIServerDeploymentFound = false
-	}
 
 	// We check whether the kube-controller-manager deployment exists in the shoot namespace. If it does not, then we assume
 	// that it has never been deployed successfully, or that we have deleted it in a previous run because we already
 	// cleaned up.
```
Comment (vpnachev) [inline pkg/controllermanager/controller/shoot/shoot_control_delete.go]: This comments should be (re)moved.

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go [hunk 11]
```diff
@@ -415,9 +445,11 @@ func (c *Controller) runDeleteShootFlow(o *operation.Operation) *gardencorev1alp
 	if err := f.Run(flow.Opts{
 		Logger:           o.Logger,
 		ProgressReporter: o.ReportShootProgress,
+		ErrorCleaner:     o.CleanShootTaskError,
+		ErrorContext:     errorContext,
 	}); err != nil {
-		o.Logger.Errorf("Error deleting Shoot %q: %+v", o.Shoot.Info.Name, err)
-		return gardencorev1alpha1helper.LastError(gardencorev1alpha1helper.FormatLastErrDescription(err), gardencorev1alpha1helper.ExtractErrorCodes(flow.Causes(err))...)
+		o.Logger.Errorf("Failed to reconcile Shoot %q: %+v", o.Shoot.Info.Name, err)
```
Comment (vpnachev) [inline pkg/controllermanager/controller/shoot/shoot_control_delete.go]: I would prefer `delete` to be used in the error message
```suggestion
		o.Logger.Errorf("Failed to delete Shoot %q: %+v", o.Shoot.Info.Name, err)
```

File: pkg/utils/reconciliationerrorhandler/error_handling.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	error
+	errorID string
+}
+
+// WithID annotates the error with the given errorID which can afterwards be retrieved by ErrorID()
+func WithID(errorID string, err error) error {
+	return &reconciliationError{err, errorID}
+}
+
+func (t *reconciliationError) ErrorID() string {
```
Comment (vpnachev) [inline pkg/utils/reconciliationerrorhandler/error_handling.go]: Needs documentation.

File: pkg/utils/reconciliationerrorhandler/error_handling.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+	return t.errorID
+}
+
+func (t *reconciliationError) Cause() error {
```
Comment (vpnachev) [inline pkg/utils/reconciliationerrorhandler/error_handling.go]: Needs documentation.

File: pkg/utils/reconciliationerrorhandler/error_handling.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+		name:         name,
+		lastErrorIDs: lastErrorIDs,
+		errorIDs:     map[string]struct{}{},
+	}
+}
+
+// AddErrorID adds an error ID which will be tracked by the context and panics if more than one error have the same ID
+func (e *ErrorContext) AddErrorID(errorID string) {
+	if e.HasErrorWithID(errorID) {
+		panic(fmt.Sprintf("Error with id %q already exists in error context %q", errorID, e.name))
```
Comment (vpnachev) [inline pkg/utils/reconciliationerrorhandler/error_handling.go]: ### Question
Can this `panic` cause the GCM process to exit with error?

**Review by vpnachev — CHANGES_REQUESTED** [review]
Thanks!

Looks very good.
I have some change requests and some questions.

File: pkg/apis/garden/v1beta1/types.go (hunk 5 above)
Comment (rfranzke) [reply pkg/apis/garden/v1beta1/types.go:753]: I don't think it's acceptable. I don't even understand why "testing it locally didn't cause errors". Can you explain that? Let's assume a shoot has a last error, the json will look like this:

`{..., "lastError: {"description": "foobar"}}`

Now, when trying to unmarshal this into `[]LastError` it will fail, right?

Comment (vpnachev) [reply pkg/apis/garden/v1beta1/types.go:753]: The field name is changed, too `lastError` -> `lastErrors`, so they are completely different fields. For sure data will be lost. But I am not sure whether it should fail if this field is presented in the etcd, but is not defined in the API?

Comment (rfranzke) [reply pkg/apis/garden/v1beta1/types.go:753]: It's simply ignored if it's in etcd but not in the API. However, wrt backwards compatibility I guess we can't do this either.

Comment (plkokanov) [reply pkg/apis/garden/v1beta1/types.go:753]: You're right that both should be present. The data loss can have an effect on the shoot care controller which I didn't take into account: if there already exist a LastError and it is not added to the LastErrors array the following code will act as if there are no errors: [shoot_care_control.go:177](https://github.com/gardener/gardener/blob/1ed7cdbbbbeb739b91fffb4f5c7d586c61a2815d/pkg/controllermanager/controller/shoot/shoot_care_control.go#L177). 
Additionally the transition to the new array field for the gardener dashboard will not have to happen immediately after it is added.

Thanks for the review and @vpnachev for pointing me to that documentation.

File: pkg/utils/reconciliationerrorhandler/error_handling.go (hunk 14 above)
Comment (plkokanov) [reply pkg/utils/reconciliationerrorhandler/error_handling.go]: The idea here is similar to the task additions in the graph which is used to generate the flow: https://github.com/gardener/gardener/blob/1ed7cdbbbbeb739b91fffb4f5c7d586c61a2815d/pkg/utils/flow/graph.go#L70

The difference is that AddErrorID is called before the execution of each task and not when ToExecute (which creates the task func) is called.

Comment (vpnachev) [reply pkg/utils/reconciliationerrorhandler/error_handling.go]: :+1:

File: pkg/apis/garden/v1beta1/types.go (hunk 5 above)
Comment (plkokanov) [reply pkg/apis/garden/v1beta1/types.go:753]: @vpnachev @rfranzke After thinking a bit about the backwards compatibility there could be the following issues:

If we keep only `LastErrors`(plural) in the internal representation of `ShootStatus` and both `LastError` and `LastErrors` in the versioned APIs: 
1. During conversion from internal to versioned to create `LastError`(singular) there is no way to construct the `LastError.Description` the same as it would have been before this change (unless we get it from the `LastOperation.Description`).
2. During updates which first retrieve the `ShootStatus` and then only change either `LastError` or `LastErrors`(plural) we have to use `LastError.LastUpdateTime` to decide whether to add `LastError` or `LastErrors`(plural) to the internal `LastErrors`. Note that `LastError.LastUpdateTime` is optional.

Comment (plkokanov) [reply pkg/apis/garden/v1beta1/types.go:753]: What I could do is to just change the LastError API by adding a new optional `TaskErrors` field:
```
type LastError {
    Description string
    // optional
    TaskID string
    // optional
    TaskErrors []TaskError
    // optional
    Codes []ErrorCodes
}

type TaskError {
   Description string
   // optional
   TaskID string
   // optional
   Codes []ErrorCodes
}
```

Two examples would be:
```
// When the error is generated during a flow execution
LastError {
    Description: Flow reconcile encountered task errors
    TaskErrors: [
        TaskError {
            Description: Faied to deploy network extension
            TaskID: Deploy network extension
        }

        TaskError {
            Description: Failed to connect to shoot apiserver
            TaskID: Connect to shoot apiserver
            Codes: "404"
        }
    ]
}

// when the error was not generated by the flow
LastError {
    Description: Failed to check versions
    TaskID: CheckVersions
}
```

When a step with a `taskID` succeeds and we have the ErrorCleaner configured:
1. Find the LastError or TaskError which has the same `taskID`.
2. If `taskID` matches the TaskID of the LastError remove it completely.
2. For the TaskErrors Create a remainingErrors array and only add TaskErrors which do not match the given `taskID` to the array
3. If len(remainingErrors) == 0 and len(LastErrors.TaskErrors) > 0 clean up LastError as well. This comparison is to avoid deleting errors which have neither TaskID nor TaskErrors.

@rfranzke @mvladev @vpnachev wdyt?

**Comment by plkokanov** [comment]
@rfranzke @vpnachev updated the PR:
**Note**: the `LastError` will always be kept as the first element in `LastErrors`(plural) from the internal representation of the `ShootStatus`. If `LastError` is set to nil then `LastErrors` will also be set to nil. Until `LastError` is removed both fields have to be maintained in the ShootStatus updates. There is no way to check if a supplied `LastError` is equal to the entire array of `LastErrors`

File: pkg/apis/core/types_common.go [hunk 15]
```diff
@@ -36,6 +36,8 @@ const (
 type LastError struct {
 	// A human readable message indicating details about the last error.
 	Description string
+	// ID of the task which caused this last error
+	TaskID *string `json:"taskID,omitempty"`
```
Comment (vpnachev) [inline pkg/apis/core/types_common.go]:
```suggestion
	TaskID *string
```

**Review by vpnachev — CHANGES_REQUESTED** [review]
a small suggestion

File: pkg/apis/core/v1alpha1/conversions.go [hunk 16]
```diff
@@ -17,6 +17,7 @@ package v1alpha1
 import (
 	"encoding/json"
 	"strings"
+	unsafe "unsafe"
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/conversions.go]:
```suggestion
	"unsafe"
```

File: pkg/apis/core/types_common.go (hunk 15 above)
Comment (plkokanov) [reply pkg/apis/core/types_common.go]: oops :)

File: pkg/apis/core/v1alpha1/conversions.go [hunk 17]
```diff
@@ -1699,6 +1710,19 @@ func Convert_v1alpha1_ShootStatus_To_garden_ShootStatus(in *ShootStatus, out *ga
 	if err := autoConvert_v1alpha1_ShootStatus_To_garden_ShootStatus(in, out, s); err != nil {
 		return err
 	}
+	if in.LastError != nil {
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/conversions.go:1715]:
```suggestion

	if in.LastError != nil {
```

File: pkg/apis/core/v1alpha1/conversions.go [hunk 18]
```diff
@@ -1699,6 +1710,19 @@ func Convert_v1alpha1_ShootStatus_To_garden_ShootStatus(in *ShootStatus, out *ga
...
+		outLastErrors := []garden.LastError{
+			garden.LastError{
+				Description:    in.LastError.Description,
+				Codes:          *(*[]garden.ErrorCode)(unsafe.Pointer(&in.LastError.Codes)),
+				LastUpdateTime: in.LastError.LastUpdateTime,
+			},
+		}
+		out.LastErrors = append(outLastErrors, *(*[]garden.LastError)(unsafe.Pointer(&in.LastErrors))...)
+
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/conversions.go]:
```suggestion
```

File: pkg/apis/garden/v1beta1/conversions.go [hunk 19]
```diff
@@ -18,7 +18,9 @@ import (
 	"encoding/json"
 	"fmt"
 	"strings"
+	unsafe "unsafe"
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/conversions.go]:
```suggestion
	"unsafe"
```

File: pkg/apis/garden/v1beta1/conversions.go [hunk 20]
```diff
@@ -2244,6 +2256,20 @@ func Convert_v1beta1_ShootStatus_To_garden_ShootStatus(in *ShootStatus, out *gar
...
+	if in.LastError != nil {
+		outLastErrors := []garden.LastError{
+			garden.LastError{
+				Description:    in.LastError.Description,
+				Codes:          *(*[]garden.ErrorCode)(unsafe.Pointer(&in.LastError.Codes)),
+				LastUpdateTime: in.LastError.LastUpdateTime,
+			},
+		}
+		out.LastErrors = append(outLastErrors, *(*[]garden.LastError)(unsafe.Pointer(&in.LastErrors))...)
+
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/conversions.go]:
```suggestion
```

File: pkg/utils/reconciliationerrorhandler/error_handling_test.go [hunk 21]
```diff
@@ -0,0 +1,160 @@
+package reconciliationerrorhandler_test
```
Comment (rfranzke) [inline pkg/utils/reconciliationerrorhandler/error_handling_test.go]: license header missing

File: pkg/utils/reconciliationerrorhandler/error_handling.go [hunk 22]
```diff
@@ -0,0 +1,140 @@
+package reconciliationerrorhandler
```
Comment (rfranzke) [inline pkg/utils/reconciliationerrorhandler/error_handling.go]: license header missing

Comment (rfranzke) [reply pkg/utils/reconciliationerrorhandler/error_handling.go]: Also, can we somehow find a better/shorter package name?

File: pkg/utils/reconciliationerrorhandler/error_handling.go [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+
+import (
+	"fmt"
+)
+
+// GetID returns the ID of the error if possible.
+// If err does not implement ErrorID or is nill an empty string will be returned.
+func GetID(err error) string {
+	type errorIDer interface {
```
Comment (rfranzke) [inline pkg/utils/reconciliationerrorhandler/error_handling.go]: a type inside a function?!

File: pkg/operation/botanist/health_check.go [hunk 24]
```diff
@@ -889,7 +889,7 @@ func isUnstableOperationType(lastOperationType gardencorev1alpha1.LastOperationT
 // in create or delete state.
 func (b *Botanist) pardonCondition(condition gardencorev1alpha1.Condition) gardencorev1alpha1.Condition {
 	shoot := b.Shoot.Info
-	if shoot.Status.LastError != nil {
+	if len(shoot.Status.LastErrors) > 0 {
```
Comment (rfranzke) [inline pkg/operation/botanist/health_check.go]: Shouldn't we also keep checking for `shoot.Status.LastError` here?

File: pkg/operation/operation.go [hunk 25]
```diff
@@ -329,6 +329,26 @@ func (o *Operation) ReportShootProgress(ctx context.Context, stats *flow.Stats)
 	o.Shoot.Info = newShoot
 }
 
+// CleanShootTaskError removes the error with taskID from the shoot status
```
Comment (rfranzke) [inline pkg/operation/operation.go]: What if the error for the given task ID is stored in `status.lastError` and not in the `status.lastErrors[]` list?

File: pkg/controllermanager/controller/shoot/shoot_utils.go [hunk 26]
```diff
@@ -114,22 +114,22 @@ func BoolToStatus(cond bool) Status {
 }
 
 // ComputeStatus computes the label Status of a shoot depending on the given lastOperation, lastError and conditions.
-func ComputeStatus(lastOperation *gardencorev1alpha1.LastOperation, lastError *gardencorev1alpha1.LastError, conditions ...gardencorev1alpha1.Condition) Status {
+func ComputeStatus(lastOperation *gardencorev1alpha1.LastOperation, lastErrors []gardencorev1alpha1.LastError, lastError *gardencorev1alpha1.LastError, conditions ...gardencorev1alpha1.Condition) Status {
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_utils.go:117]: What about `.status.lastError`?

File: pkg/controllermanager/controller/shoot/shoot_control_reconcile.go [hunk 27]
```diff
@@ -411,15 +436,21 @@ func (c *Controller) updateShootStatusReconcileSuccess(o *operation.Operation, o
...
 		progress      = 1
 		willRetry     = !utils.TimeElapsed(o.Shoot.Info.Status.RetryCycleStartTime, c.config.Controllers.Shoot.RetryDuration.Duration)
 	)
 
+	// TODO: Remove this after LastError is removed from the ShootStatus API
+	var codes []gardencorev1alpha1.ErrorCode
+	for _, lastErr := range lastErrors {
+		codes = append(codes, lastErr.Codes...)
+	}
+	lastError := gardencorev1alpha1helper.LastError(description, codes...)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_reconcile.go:446]: Okay, so `status.lastError` contains all errors, and `status.lastErrors[]` contains also all errors, however, separated into a list with dedicated entries, right?

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go (hunk 9 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/shoot/shoot_control_delete.go]: Yes, moved above, please.

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go (hunk 10 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/shoot/shoot_control_delete.go]: Yes, moved above, please.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/utils/reconciliationerrorhandler/error_handling.go (hunk 22 above)
Comment (plkokanov) [reply pkg/utils/reconciliationerrorhandler/error_handling.go]: will try to come up with something shorter.

File: pkg/controllermanager/controller/shoot/shoot_control_reconcile.go (hunk 27 above)
Comment (plkokanov) [reply pkg/controllermanager/controller/shoot/shoot_control_reconcile.go]: I've seen this used in [this errors package][1] and some other places. 
The idea is that you force extracting the `id` only from the `GetID(err error)` method.

In our case this might be a bit too much limitation as there isn't any special logic (as in the cycle in [the error package][1])

[1]: https://github.com/pkg/errors/blob/master/errors.go#L270

File: pkg/utils/reconciliationerrorhandler/error_handling.go (hunk 23 above)
Comment (plkokanov) [reply pkg/utils/reconciliationerrorhandler/error_handling.go]: I've seen this used in [this errors package][1] and some other places. 
The idea is that you force extracting the `id` only from the `GetID(err error)` method.

In our case this might be a bit too much limitation as there isn't any special logic (as in the for loop in [the error package][1])

[1]: https://github.com/pkg/errors/blob/master/errors.go#L270

File: pkg/operation/botanist/health_check.go (hunk 24 above)
Comment (plkokanov) [reply pkg/operation/botanist/health_check.go]: Yep, missed it here.

File: pkg/controllermanager/controller/shoot/shoot_utils.go (hunk 26 above)
Comment (plkokanov) [reply pkg/controllermanager/controller/shoot/shoot_utils.go]: Hmm what do you mean? it's right after `lastErrors []gardencorev1alpha1.LastError,`

File: pkg/controllermanager/controller/shoot/shoot_control_reconcile.go (hunk 27 above)
Comment (plkokanov) [reply pkg/controllermanager/controller/shoot/shoot_control_reconcile.go:446]: Yes, but there is one difference.  The error descriptions in status.lastErrors[].Descriptions do not contain the following text:
`flow %q encountered task errors` which is present in status.LastError.Description and status.LastOperation.Description

File: pkg/apis/core/v1alpha1/helper/errors.go [hunk 28]
```diff
@@ -22,6 +22,8 @@ import (
 
 	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
 	"github.com/gardener/gardener/pkg/utils"
+	"github.com/gardener/gardener/pkg/utils/reconciliationerrorhandler"
+	errors2 "github.com/pkg/errors"
```
Comment (zanetworker) [inline pkg/apis/core/v1alpha1/helper/errors.go:26]: can you rename this package to something else? also do we really need it?

File: pkg/apis/core/v1alpha1/conversions.go [hunk 29]
```diff
@@ -1699,6 +1710,19 @@ func Convert_v1alpha1_ShootStatus_To_garden_ShootStatus(in *ShootStatus, out *ga
 	if err := autoConvert_v1alpha1_ShootStatus_To_garden_ShootStatus(in, out, s); err != nil {
 		return err
 	}
+	if in.LastError != nil {
+		outLastErrors := []garden.LastError{
+			garden.LastError{
```
Comment (zanetworker) [inline pkg/apis/core/v1alpha1/conversions.go]: you don't need to write `garden.LastError here`
```suggestion
			{
```

File: pkg/apis/garden/v1beta1/conversions.go [hunk 30]
```diff
@@ -2244,6 +2256,20 @@ func Convert_v1beta1_ShootStatus_To_garden_ShootStatus(in *ShootStatus, out *gar
 		return err
 	}
 
+	if in.LastError != nil {
+		outLastErrors := []garden.LastError{
+			garden.LastError{
```
Comment (zanetworker) [inline pkg/apis/garden/v1beta1/conversions.go]: you don't need to write garden.LastError here. 
```suggestion
			{
```

File: pkg/operation/operation.go (hunk 25 above)
Comment (plkokanov) [reply pkg/operation/operation.go]: Currently status.LastError doesn't have TaskIDs added to it during reconciliation. 
I can configure the logic in CleanShootTaskError to also check the TaskID of status.LastError. 

But there could be a minor inconvenience if someone adds TaskID to a status.LastError which is comprised of multiple errors -> errors on steps that haven't yet completed could be cleared.

File: pkg/apis/core/v1alpha1/helper/errors.go (hunk 28 above)
Comment (plkokanov) [reply pkg/apis/core/v1alpha1/helper/errors.go:26]: Moved the code in `pkg/utils/errors`.

File: pkg/operation/operation.go (hunk 25 above)
Comment (rfranzke) [reply pkg/operation/operation.go]: OK, I see, thank you for the explanation!

File: pkg/utils/errors/errors_test.go [hunk 31]
```diff
@@ -1,9 +1,13 @@
-package errors
+package errors_test
```
Comment (rfranzke) [inline pkg/utils/errors/errors_test.go:15]: add the license header please

**Review by rfranzke — COMMENTED** [review]
Thanks for updating, I have no further comments on that point in time, I need to try it out in practice, still.
@vpnachev anything from you?

**Comment by rfranzke** [comment]
Could you please rebase and squash appropriately?

**Comment by plkokanov** [comment]
squashed and rebased

**Review by vpnachev — APPROVED** [review]
/approve

**Review by rfranzke — APPROVED** [review]

**Comment by rfranzke** [comment]
@plkokanov can you refine the release note to explain that there is now a new `lastErrors` field and how it is different to `lastError`?
